source/Irrlicht/COpenGLDriver.h

```diff
@@ -104,43 +110,52 @@ namespace video
             GLboolean depthWriteEnable = 1;
             //GLboolean depthBoundsTestEnable;
             GLboolean stencilTestEnable = 0;
+
+            GLboolean logicOpEnable = 0;
+            GLenum logicOp = GL_COPY;
+            struct SDrawbufferBlending
+            {
+                GLboolean blendEnable = 0;
+                struct SBlendFunc {
+                    GLenum srcRGB = GL_ONE;
+                    GLenum dstRGB = GL_ZERO;
+                    GLenum srcAlpha = GL_ONE;
+                    GLenum dstAlpha = GL_ZERO;
+                    bool operator!=(const SBlendFunc& rhs) const { return srcRGB!=rhs.srcRGB || dstRGB!=rhs.dstRGB || srcAlpha!=rhs.srcAlpha || dstAlpha!=rhs.dstAlpha; }
+                } blendFunc;
+                struct SBlendEq {
+                    GLenum modeRGB = GL_FUNC_ADD;
+                    GLenum modeAlpha = GL_FUNC_ADD;
+                    bool operator!=(const SBlendEq& rhs) const { return modeRGB!=rhs.modeRGB || modeAlpha!=rhs.modeAlpha; }
+                } blendEquation;
+                struct SColorWritemask {
+                    GLboolean colorWritemask[4]{ 1,1,1,1 };
+                    bool operator!=(const SColorWritemask& rhs) const { return memcmp(colorWritemask, rhs.colorWritemask, 4); }
+                } colorMask;
+            } drawbufferBlend[asset::SBlendParams::MAX_COLOR_ATTACHMENT_COUNT];
         } rasterParams;
 
         struct {
-            //GLuint vao; //vao will be found in cache (or created) based on params
-            uint32_t mapAttrToBinding[16]{};
-            struct {
+            HashVAOPair vao;
+            struct SBnd {
                 GLuint buf = 0;
                 GLintptr offset = 0;
-                GLsizei stride = 0;
+                bool operator!=(const SBnd& rhs) const { return buf!=rhs.buf || offset!=rhs.offset; }
             } bindings[16];
-            uint16_t divisors = 0x0000;
             GLuint indexBuf;
         } vertexInputParams;
 
         struct {
-            struct SBuffers {
-                GLuint buffers[16]{};
-                GLintptr offsets[16]{};
-                GLsizeiptr sizes[16]{};
-            };
-            SBuffers ubos;
-            SBuffers ssbos;
-
-            struct {
-                GLuint textures[80]{};
-                GLenum targets[80];
-            } textures;
-            GLuint samplers[80]{};
-            struct {
-                GLuint textures[80]{};
-                GLenum formats[80];
-            } images;
+            const COpenGLDescriptorSet* descSets[4];

```

Comment:
smart refcounted pointer, you dont want the system to go tits up because it was dropped because reference was not being held onto while its still set as current or next

Reply:
resolved